README.md
# VS Code external-app-debug Extension

This is an extension designed to facilitate development of content that is executed inside of external applications, by providing integration with debug-protocol-aware applications

## Using the external-app-debug extension

* Install the **external-app-debug** extension in VS Code
* In VS Code's Preferences, under "Extensions > External-App Debug", set the application you want to launch
* In the Debug viewlet, click the dropdown to "Add Configuration..." and select "External-App Debug" to create a debug configuration in your launch.json
	* Note that `launch` and `attach` request types are available in the extension
	* Other settings such as `stopAtEntry` and `program` must be acknowledged by the target application

From there, VS Code should launch the  you provided, and establish a connection with it. Functionality from that point forward is handled by the application, including systems such as loading of symbols, controlling execution, fetching runtime state, and so on.

An example of a supported application (itself a WIP) is available at https://github.com/CypherSignal/bsnes-plus/tree/vscode-newdbg

## Information for application authors

If you are the maintainer of an application and you want it to work with this extension as a debug target, note the following:

When running the `launch` request, your application will be launched as a new process, and the extension will do a portscan from ports 5420 to 5450 looking for the process in question, by attempting to create Tcp Connections at each port. Your application should run a Tcp Server in this port range, listening for connections. If the extension successfully makes a connection with your application, a custom Request, with the command `external-apppreinit`, is sent, similar to normal Requests fired as part of the Debug Adapter Protocol. A Response must be provided, filling out your application's `pid` (Process ID). Once your application's Response is received, the connection that was created for the portscan will be closed, and the port your application was discovered on will be passed to VS Code. VS Code which will then open a new connection on the same port that your process was discovered on, and perform normal Debug Adapter Protocol activities.

When running the `attach` request, some similar behaviour will occur. Of course, your application will not be launched, but a portscan over the same set of ports as above will be performed. the `external-apppreinit` request will also be fired, but in addition to your application's pid, you should also fill out the `title` and `description` of the body. The extension will enumerate all respondees to the `external-apppreinit` request, and display the `title` and `description` for the user in a Picker. It is recommended for the `title` to be your application's name, and `description` to be the name of the script or sub-program currently loaded, if any.

If the `external-apppreinit` request was included in the Debug Adapter Protocol, it would resemble this definition:

```
export interface external-appPreInitRequest extends Request {
 	// command: 'external-apppreinit';
}

export interface external-appPreInitResponse extends Response {
	// Information for this debug adapter
	body: {
		pid: Number; // The process ID of the executable
		title: String; // The title to display on the picker, for selecting executable to attach to
		description: String; // An extended description of what to attach to
	}
}
```

Aside from that, all interactions with the Debug Adapter Protocol are unchanged. Information on this protocol can be found at https://microsoft.github.io/debug-adapter-protocol/overview.


## Build and run the extension yourself

* Clone the project, and open the project folder in VS Code
* Run `npm install` in the terminal
* Run the "Extension" debug configuration


CHANGELOG.md
## 0.5.0
* Dependency and security update.

## 0.4.0
* Dependency and security update.

## 0.3.0
* Dependency and security update.

## 0.2.0
* Fix issue with socket cleanup during launch/attach not being thorough enough.

## 0.1.0
* Initial release, allowing for a user to specify and launch an external app in question.



src/extension.ts
import * as net from 'net';
import * as child_process from 'child_process';
import * as vscode from 'vscode';
import { WorkspaceFolder, DebugConfiguration, CancellationToken, ProviderResult } from 'vscode';

export function activate(context: vscode.ExtensionContext) {
	context.subscriptions.push(vscode.commands.registerCommand('extension.external-app-debug.getFileLocation', config => {
		return vscode.window.showInputBox({
			placeHolder: "Specify the name of an file in the workspace folder for the application to execute"
		});
	}));

	// register a configuration provider for external-app-debug
	const provider = new ExternalAppDebugConfigurationProvider();
	context.subscriptions.push(vscode.debug.registerDebugConfigurationProvider('external-app-debug', provider));
	context.subscriptions.push(provider);

	// register an adapter descriptor for external-app-debug
	const factory = new ExternalAppDebugAdapterDescriptorFactory();
	context.subscriptions.push(vscode.debug.registerDebugAdapterDescriptorFactory('external-app-debug', factory));
	context.subscriptions.push(factory);
}

class PortItem implements vscode.QuickPickItem {
	port: number;
	pid: number;
	label: string;
	description: string;
	constructor(port:number, pid = 0, label = "", description = "") {
		this.port = port;
		this.pid = pid;
		this.label = label;
		this.description = description;
	}
}

function getExternalAppLocation() : string {
	return vscode.workspace.getConfiguration('external-app-debug').get<string>('externalAppPath') ||
			(vscode.window.showErrorMessage("external-app-debug externalAppPath setting not configured."), "");
}

// generate an array of promises that each (attempt) to establish a connection on the given port,
// and once connection is made, send down a custom "debug-adapter"-style json request
// then wait for the response to fill out the data
//
// Sample version of the ExternalAppDebugPreInit Request:
//
// export interface ExternalAppDebugPreInitRequest extends Request {
// 	// command: 'external-apppreinit';
// }
//
// Sample version of the ExternalAppDebugPreInit response that should be filled out by the target executable
//
// 	export interface ExternalAppDebugPreInitResponse extends Response {
// 		// Information for this debug adapter
// 		body: {
// 			pid: Number; // The process ID of the executable
// 			title: String; // The title to display on the picker, for selecting executable to attach to
// 			description: String; // An extended description of what to attach to
//	 	}
// 	}
async function getListeningProcesses() : Promise<PortItem[]> {
	let portscanResults = new Array<Promise<PortItem | undefined>>();
	for (let i = 5420; i < 5450; i++) {
		portscanResults.push(new Promise((resolve, reject) => {
			let socket = net.createConnection({port : i}, () => {
					// after creating the connection, send a "preinit" request message
					// and get the response to fill out the PortItem data
					const request: any = {
						command: "external-apppreinit",
						type: "request",
						seq: 0
					};
					const requestJson = JSON.stringify(request);
					socket.write(`Content-Length: ${Buffer.byteLength(requestJson, 'utf8')}\r\n\r\n${requestJson}`, 'utf8');
				})
				.on('data', (data: Buffer) => {
					const lines = data.toString('utf8').split(/\r?\n/);
					if (lines.length === 4) {
						const content = JSON.parse(lines[2]);
						if (content.command === "external-apppreinit") {
							if (content.body !== undefined) {
								resolve(new PortItem(i, content.body.pid, (content.body.title || `Port: ${i}`) + ` | PID: ${content.body.pid}`, content.body.description));
							}
							else {
								resolve(new PortItem(i));
							}
							socket.end();
							return;
						}
					}
					resolve(undefined);
					socket.end();
				})
				.on('error', () => { resolve(undefined); socket.destroy(); })
				.on('timeout', () => { resolve(undefined); socket.destroy(); })
				.setTimeout(150);
		}));
	}
	return Promise.all(portscanResults).then(results => results.filter(item => item !== undefined) as PortItem[]);
}

// Repeatedly does a portscan for an executable matching the given ProcessId for a few seconds
async function getListenerPortForProcess(pid:number) : Promise<PortItem | undefined> {
	const endTimeout = new Date().getTime() + 5000;
	while (new Date().getTime() < endTimeout) {
		const resultPromise = await getListeningProcesses().then(items => items.find(item => item.pid === pid));
		if (resultPromise) {
			return resultPromise;
		}
		else { // do a short wait to we don't spam constantly
			await new Promise( _ => setTimeout(_, 100));
		}
	}
	return undefined;
}

class ExternalAppDebugConfigurationProvider implements vscode.DebugConfigurationProvider {

	// Massage a debug configuration just before a debug session is being launched,
	// e.g. throw out the program name if we're doing an attach
	resolveDebugConfiguration(folder: WorkspaceFolder | undefined, config: DebugConfiguration, token?: CancellationToken): ProviderResult<DebugConfiguration> {
		if (config.request === "attach") {
			config.program = "";
		}
		return config;
	}

	dispose() {	}
}

// factory that generates the DebugAdapterServer after finalizing the debug configuration
class ExternalAppDebugAdapterDescriptorFactory implements vscode.DebugAdapterDescriptorFactory  {

	createDebugAdapterDescriptor(session: vscode.DebugSession, executable: vscode.DebugAdapterExecutable | undefined): vscode.ProviderResult<vscode.DebugAdapterDescriptor> {
		let portPromise:Thenable<PortItem | undefined>;

		if (session.configuration.request === "launch") {
			// launch the process to attach to
			const process = child_process.spawn(getExternalAppLocation());

			// start portscan to find the process to connect to, and do so
			portPromise = getListenerPortForProcess(process.pid)
		}
		else if (session.configuration.request === "attach") {
			// do a portscan and use that to populate a picker
			portPromise = vscode.window.showQuickPick(getListeningProcesses());
		}
		else {
			portPromise = new Promise(() => { return undefined; });
		}

		// convert the promised PortItem into a debugAdapterServer
		return portPromise.then(result => {
			if (result) {
				return new vscode.DebugAdapterServer(result.port);
			}
			return undefined;
		});
	}

	dispose() {	}
}